### PR TITLE
add the state parameter to the oidc authentication request

### PR DESCRIPTION
**Pull request by mortbauer**
* added state paramter to make resilient against CSRF attacks
* see issue #3465 for more information

## Related Issues
 * #3465

### CODE REVIEW

File: owncloudApp/src/main/java/com/owncloud/android/presentation/ui/authentication/LoginActivity.kt [hunk 1]
```diff
@@ -445,13 +448,15 @@ class LoginActivity : AppCompatActivity(), SslUntrustedCertDialog.OnSslUntrusted
         val customTabsBuilder: CustomTabsIntent.Builder = CustomTabsIntent.Builder()
         val customTabsIntent: CustomTabsIntent = customTabsBuilder.build()
 
+        this.oidcState = UUID.randomUUID().toString().substring(0,15)
```
Comment (IljaN) [inline owncloudApp/src/main/java/com/owncloud/android/presentation/ui/authentication/LoginActivity.kt:451]: Isn't it better to use base64 encoded bytes from SecureRandom here?

Comment (mortbauer) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/ui/authentication/LoginActivity.kt]: definitely better, created a new function in OauthUtils for that

File: owncloudApp/src/main/java/com/owncloud/android/presentation/ui/authentication/LoginActivity.kt [hunk 2]
```diff
@@ -445,13 +447,15 @@ class LoginActivity : AppCompatActivity(), SslUntrustedCertDialog.OnSslUntrusted
...
+        this.oidcState = OAuthUtils().generateRandomState()
         val authorizationEndpointUri = OAuthUtils.buildAuthorizationRequest(
             authorizationEndpoint = authorizationEndpoint,
             redirectUri = OAuthUtils.buildRedirectUri(applicationContext).toString(),
             clientId = clientId,
             responseType = ResponseType.CODE.string,
             scope = if (oidcSupported) OAUTH2_OIDC_SCOPE else "",
-            codeChallenge = oauthViewModel.codeChallenge
+            codeChallenge = oauthViewModel.codeChallenge,
+            state = this.oidcState
```
Comment (abelgardep) [inline owncloudApp/src/main/java/com/owncloud/android/presentation/ui/authentication/LoginActivity.kt:458]: I think that we could generate the state in the ViewModel as we do with `codeChallenge`. What do you think?

https://github.com/owncloud/android/blob/master/owncloudApp/src/main/java/com/owncloud/android/presentation/viewmodels/oauth/OAuthViewModel.kt#L47

**Review by abelgardep — CHANGES_REQUESTED** [review]
Just a tiny change and we are ready to go @mortbauer

BTW, rebase this branch with master when you can 👍

File: owncloudApp/src/main/java/com/owncloud/android/presentation/ui/authentication/LoginActivity.kt (hunk 2 above)
Comment (mortbauer) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/ui/authentication/LoginActivity.kt]: Please have a look again, I just pushed the minor change, I hope this is how you meant it, at least it makes sense to me, looks more consistent now.

**Review by abelgardep — APPROVED** [review]
Thanks for the contribution!!

LGTM, @jesmrec ready to QA

**Comment by jesmrec** [comment]
QA checks, regression from current authentication methods

- [X] OIDC + DCR (test instance by jw)
- [X] OIDC without DCR (`ocis.owncloud.works`) 
- [x] OAuth2 session creation (regression)
- [x] OAuth2 token renewal (regression)
- [x] Basic auth (regression)

Approved from my side

**Pull request merged**